## 收集

In [2]:
import pandas as pd

In [3]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## 评估

In [4]:
patients

patient_id assigned_sex  given_name       surname  \
0             1       female         Zoe       Wellish   
1             2       female      Pamela          Hill   
2             3         male         Jae        Debord   
3             4         male        Liêm          Phan   
4             5         male         Tim       Neudorf   
5             6         male      Rafael         Costa   
6             7       female        Mary         Adams   
7             8       female      Xiuxiu         Chang   
8             9         male       Dsvid    Gustafsson   
9            10       female      Sophie       Cabrera   
10           11       female       Sandy    Gunnarsson   
11           12         male   Abdul-Nur           Isa   
12           13         male  Omeokachie      Ibeamaka   
13           14       female    Anenechi         Chidi   
14           15       female        Asia       Woźniak   
15           16         male       Søren          Lund   
16           17       female         Tám          Liễu   
17           18       female     Roxanne     Andreyeva   
18           19         male     William         Oates   
19           20         male         Zak         Kelly   
20           21       female       Sofia       Karlsen   
21           22         male      Samúel  Guðbrandsson   
22           23         male      Manchu            Su   
23           24         male       Lovre         Galić   
24           25         male       Jakob      Jakobsen   
25           26         male      Gregor          Bole   
26           27       female        Ella          Lund   
27           28         male      Joseph        Tucker   
28           29         male      Robert          Wolf   
29           30         male        Jake      Jakobsen   
..          ...          ...         ...           ...   
473         474       female        Kate     Wilkinson   
474         475       female   Esperanza      Labrosse   
475         476         male       Malik       Vaneker   
476         477       female       Berta    Napolitani   
477         478         male     Juliusz      Majewski   
478         479       female      Edelma   Villalpando   
479         480         male        Tapa   Arsanukayev   
480         481         male      Nasser       Mansour   
481         482         male     Michael    Kristensen   
482         483         male       Diogo         Souza   
483         484       female       Angel         Grant   
484         485         male     Placido       Udinesi   
485         486         male      Trifon      Izmailov   
486         487         male      Samuel          Blix   
487         488         male        Ivar       Löfgren   
488         489         male        Mika    Martinsson   
489         490       female     Jasmine         Sykes   
490         491         male     Jackson       Addison   
491         492       female     Vanessa      Ferguson   
492         493         male       Poldi           Tar   
493         494       female         Fen          Chin   
494         495       female      Sirkka     Piirainen   
495         496         male      Hajime       Tsukada   
496         497         male   Alexander        Hueber   
497         498         male    Masataka      Murakami   
498         499         male     Mustafa     Lindström   
499         500         male       Ruman       Bisliev   
500         501       female       Jinke     de Keizer   
501         502       female     Chidalu   Onyekaozulu   
502         503         male         Pat       Gersten   

                        address               city       state  zip_code  \
0          576 Brown Bear Drive  Rancho California  California   92390.0   
1     2370 University Hill Road          Armstrong    Illinois   61812.0   
2         1493 Poling Farm Road               York    Nebraska   68467.0   
3           2335 Webster Street         Woodbridge          NJ    7095.0   
4          1428 Turkey Pen L

In [5]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
250         chen           yao          -  56u - 57u         7.90       7.51   
251        aksel   vestergaard          -  42u - 38u         9.62       9.29   
252        ellen         luman          -  40u - 39u         9.27       8.77   
253       albino     schiavone  35u - 43u          -         7.56       7.15   
254         jose         combs          -  39u - 36u         7.89       7.42   
255       jia li          teng  48u - 54u          -         7.66       7.32   
256        ilija        horvat  42u - 50u          -         7.77       7.38   
257     mathilde      nørgaard          -  27u - 28u         8.50       8.10   
258       csilla      herczegh          -  43u - 46u         7.71       7.27   
259      aaliyah          rice          -  31u - 31u         7.64       7.33   
260        david      beauvais          -  26u - 23u         7.87       7.47   
261     caroline        shuler          -  50u - 54u         7.63       7.27   
262         alex      crawford  51u - 62u          -         7.69       7.30   
263      rebecca      jephcott  53u - 63u          -         7.96       7.57   
264   chukwumoge     ogochukwu          -  41u - 39u         7.95       7.56   
265       fearne      mcgregor          -  27u - 29u         7.83       7.48   
266       ursula         freud  42u - 54u          -         7.75       7.46   
267         leon        scholz          -  38u - 32u         7.72       7.29   
268   

In [6]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [7]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [8]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [9]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [10]:
list(adverse_reactions)

['given_name', 'surname', 'adverse_reaction']

In [11]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [12]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [13]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [14]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [15]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [16]:
patients.sample(5)

patient_id assigned_sex given_name   surname               address  \
425         426       female     Taylah  Mobourne  995 Beechwood Avenue   
252         253         male       Juan     Pérez  2549 Pearlman Avenue   
199         200         male     Zdeněk     Synek  3818 Strother Street   
170         171         male    Chikere    Achebe   321 Briercliff Road   
206         207       female       Leah      Wong    3548 Hilltop Drive   

          city     state  zip_code        country  \
425  Weehawken        NJ    7087.0  United States   
252    Bedford        MA    1730.0  United States   
199       Dora        AL   35062.0  United States   
170   Brooklyn  New York   11227.0  United States   
206   Canadian        TX   79014.0  United States   

                                      contact  birthdate  weight  height   bmi  
425      908-751-4255TaylahMobourne@rhyta.com   7/4/1961   188.5      67  29.5  
252         978-243-8596JuanPerez@teleworm.us  10/1/1991   224.2      70  32.2  
199  ZdenekSynek@jourrapide.com1 205 648 4197  9/27/1950   168.3      70  24.1  
170  718-628-9500ChikereAchebe@jourrapide.com  1/12/1939   215.4      67  33.7  
206           806-323-2210LeahWong@einrot.com  6/11/1934   129.8      62  23.7

In [17]:
patients.surname.value_counts()

Doe            6
Jakobsen       3
Taylor         3
Correia        2
Hueber         2
Berg           2
Kadyrov        2
Johnson        2
Ogochukwu      2
Collins        2
Nilsen         2
Kowalczyk      2
Cabrera        2
Silva          2
Bùi            2
Schiavone      2
Souza          2
Grímsdóttir    2
Tạ             2
Lâm            2
Liễu           2
Woźniak        2
Gersten        2
Lương          2
Dratchev       2
Parker         2
Tucker         2
Lund           2
Batukayev      2
Cindrić        2
              ..
Barsukov       1
Richardson     1
Baer           1
Shishani       1
Csorba         1
Luwam          1
Reilly         1
Fejes          1
Crawford       1
Zsinkó         1
Koffeman       1
Kelly          1
Hsueh          1
Filípek        1
Piirainen      1
Salib          1
Chin           1
Eskelinen      1
Pecha          1
Takenaka       1
Resanovič      1
Webb           1
Eldarkhanov    1
Hsu            1
Sleiman        1
Novosel        1
Freud          1
Chung         

In [18]:
patients.address.value_counts()

123 Main Street               6
2778 North Avenue             2
648 Old Dear Lane             2
2476 Fulton Street            2
312 Jim Rosa Lane             1
4110 Capitol Avenue           1
1435 Limer Street             1
3006 Maple Court              1
2931 Romano Street            1
909 Williams Avenue           1
2386 Linda Street             1
3303 Anmoore Road             1
2172 Lynn Street              1
4720 Gordon Street            1
959 Heritage Road             1
4033 White Avenue             1
4334 Black Oak Hollow Road    1
2915 Lynn Avenue              1
2945 Ferguson Street          1
4476 Center Street            1
4992 Moonlight Drive          1
1116 Dog Hill Lane            1
4273 Deans Lane               1
2043 Jadewood Drive           1
1786 Gerald L. Bates Drive    1
4386 Camden Street            1
3479 Elm Drive                1
1975 Holden Street            1
936 Lightning Point Drive     1
3688 Adonais Way              1
                             ..
962 Geor

In [19]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [20]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [21]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [22]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [23]:
sum(treatments.auralin.isnull())

0

In [24]:
sum(treatments.novodra.isnull())

0

#### 质量
##### `patients` 表格
- 邮编是浮点，而不是字符串
- 邮编有时是四位数
- Tim Neudorf 的身高是 27 英尺，而不是 72 英尺
- 有时使用州的全称，其他用简称
- Dsvid Gustafsson
- 缺少人口统计信息 (地址列至联系方式列) ***(无法清理)***
- 错误的数据类型 (指定性别、州、邮编和出生日期列)
- 多个手机号格式
- John Doe 的默认数据
- Jakobsen、Gersten、Taylor 多条记录
- Zaitseva 体重单位是 kgs，不是 lbs 

##### `treatments` 表格
- 缺少 HbA1c 变化
- Auralin 和 Novodra 初始剂量和最终剂量中的字母 'u' 
- 小写形式的姓和名
- 缺少记录 (280 而不是 350)
- 错误的数据类型 (auralin 和 novodra 列)
- HbA1c 变化不准确 (多个开头的 4 被误写为 9 )
- auralin 和 novodra 列中空值用破折号 (-) 表示

##### `adverse_reactions` 表格
- 小写形式的姓和名

#### 清洁度
-  `patients` 表格中的联系方式一列应该拆分为手机号和邮箱两列
-  `treatments` 表格两列中的三个变量 (治疗、初始剂量和最终剂量)
-  不良反应应是 `treatments` 表格的一部分
-  `patients` 表格中的姓和名两列复制到 `treatments` 和 `adverse_reactions` 表格中

## 清理

In [25]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### 缺失数据

<font color='red'>观看 *"先处理数据缺失问题"* 视频后，完成下列两个 "缺失数据" **定义、代码和测试** 顺序。</font>

#### `treatments`： 缺少记录 (280 而不是 350)

##### 定义
*把缺少的数据用concat命令拼接起来，用ignore_index使索引不会重复。注释：缺少的 `treatments` 记录存储在文件中，命名为 `treatments_cut.csv`，你可以在这个 Jupyter Notebook 指示板中查看 (点击 Notebook 左上角的 **jupyter** 标志)。提示： 解答中使用的函数 [文档页面](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) 。*

##### 代码

In [26]:
# 这里是你的清理代码
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_clean = pd.concat([treatments_clean, treatments_cut], ignore_index=True)

##### 测试

In [27]:
# 这里是你的测试代码
treatments_clean.sample(10)

given_name      surname    auralin    novodra  hba1c_start  hba1c_end  \
213    cecilie       nilsen          -  49u - 50u         9.63       9.21   
315      alvin      jackson  38u - 43u          -         7.62       7.23   
217     taylah     mobourne  41u - 50u          -         7.90       7.51   
32       laura  ehrlichmann          -  43u - 40u         7.95       7.46   
100        kai        cunha          -  37u - 36u         7.81       7.42   
309      luboš        pecha          -  30u - 27u         7.79       7.45   
124       nate       bowden          -  48u - 47u         9.81       9.30   
307        zoe      wellish          -  33u - 33u         7.71       7.30   
40        ásta  grímsdóttir          -  29u - 30u         7.62       7.16   
201       lena         baer          -  41u - 38u         7.70       7.40   

     hba1c_change  
213          0.92  
315          0.39  
217          0.39  
32           0.99  
100          0.39  
309          0.34  
124          0.51  
307          0.91  
40           0.96  
201          0.30

#### `treatments`：缺少 HbA1c 变化和 HbA1c 变化不准确 (多个开头的 4 被误写为 9)
*注释： "HbA1c 变化不准确 (多个开头的 4 被误写为 9)" 的观察属于准确度问题，而不是完整性问题，由于通过修改缺失的 "缺少 HbA1c 变化" 的清理操作，因此这也被包含在这个标题下面。在一个 **定义、代码和测试** 表头中多个观察数据会在这个记事本中出现。*

##### 定义
*用HbA1c前后的变化相减来得到准确的值*

##### 代码

In [28]:
# 这里是你的清理代码
treatments_clean['hba1c_change'] = treatments_clean['hba1c_start'] - treatments_clean['hba1c_end']

##### 测试

In [29]:
# 这里是你的测试代码
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

### 清洁度

<font color='red'>观看 *"清理之整洁度"* 视频后，完成下列四个 "清洁度" **定义、代码和测试** 顺序。</font>

####  `patients` 表格中的联系方式一列应该包括两个变量：手机号和邮箱

##### 定义
*把联系方式分成两列，用extract的捕获正则表达式中的组，提示 1：通过 pandas 的 [`str.extract` 方法](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html) 使用常规表达。这是个神奇的 [正则表达式教程](https://regexone.com/)。提示 2： [各种手机号正则表达式模式](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number)。提示 3：[邮箱地址正则表达式模式](http://emailregex.com/)，你可能需要用这个进行修改，区分邮箱地址和手机号。*

##### 代码

In [30]:
patients_clean.contact.head()

0          951-719-9170ZoeWellish@superrito.com
1         PamelaSHill@cuvox.de+1 (217) 569-3204
2              402-363-6804JaeMDebord@gustr.com
3    PhanBaLiem@jourrapide.com+1 (732) 636-8246
4               334-515-7487TimNeudorf@cuvox.de
Name: contact, dtype: object

In [31]:
# 这里是你的清理代码
email = patients_clean.contact.str.extract(r'(?P<email>[a-zA-Z_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z-.]+)')
phone = patients_clean.contact.str.extract(r'(?P<phone>(\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]\d{3}[\s.-]\d{4})')

In [32]:
patients_clean.drop('contact', axis=1, inplace=True)

In [33]:
patients_clean = patients_clean.assign(email=email.email,phone = phone.phone)

##### 测试

In [78]:
# 这里是你的测试代码
patients_clean[patients_clean.phone.isnull()]

patient_id assigned_sex given_name      surname          address  \
209         210       female     Lalita  Eldarkhanov              NaN   
215         216         male       John          Doe  123 Main Street   
219         220         male         Mỹ        Quynh              NaN   
229         230         male       John          Doe  123 Main Street   
230         231       female  Elisabeth      Knudsen              NaN   
234         235       female    Martina    Tománková              NaN   
237         238         male       John          Doe  123 Main Street   
242         243         male       John      O'Brian              NaN   
244         245         male       John          Doe  123 Main Street   
249         250         male   Benjamin       Mehler              NaN   
251         252         male       John          Doe  123 Main Street   
257         258         male        Jin         Kung              NaN   
264         265       female   Wafiyyah       Asfour              NaN   
269         270       female     Flavia   Fiorentino              NaN   
277         278         male       John          Doe  123 Main Street   
278         279       female   Generosa        Cabán              NaN   
286         287         male      Lewis         Webb              NaN   
296         297       female        Chỉ          Lâm              NaN   

         city state zip_code        country  birthdate  weight  height   bmi  \
209       NaN   NaN    00000            NaN 1950-08-14   143.4      62  26.2   
215  New York    NY    12345  United States 1975-01-01   180.0      72  24.4   
219       NaN   NaN    00000            NaN 1978-04-09   237.8      69  35.1   
229  New York    NY    12345  United States 1975-01-01   180.0      72  24.4   
230       NaN   NaN    00000            NaN 1976-09-23   165.9      63  29.4   
234       NaN   NaN    00000            NaN 1936-04-07   199.5      65  33.2   
237  New York    NY    12345  United States 1975-01-01   180.0      72  24.4   
242       NaN   NaN    00000            NaN 1957-02-25   205.3      74  26.4   
244  New York    NY    12345  United States 1975-01-01   180.0      72  24.4   
249       NaN   NaN    00000            NaN 1951-10-30   146.5      69  21.6   
251  New York    NY    12345  United States 1975-01-01   180.0      72  24.4   
257       NaN   NaN    00000            NaN 1995-05-17   231.7      69  34.2   
264       NaN   NaN    00000            NaN 1989-11-03   158.6      63  28.1   
269       NaN   NaN    00000            NaN 1937-10-09   175.2      61  33.1   
277  New York    NY    12345  United States 1975-01-01   180.0      72  24.4   
278       NaN   NaN    00000            NaN 1962-12-16   124.3      69  18.4   
286       NaN   NaN    00000            NaN 1979-04-01   155.3      68  23.6   
296       NaN   NaN    00000            NaN 1990-05-14   181.1      63  32.1   

                 email phone  
209                NaN   NaN  
215  johndoe@email.com   NaN  
219                NaN   NaN  
229  johndoe@email.com   NaN  
230                NaN   NaN  
234                NaN   NaN  
237  johndoe@email.com   NaN  
242                NaN   NaN  
244  johndoe@email.com   NaN  
249                NaN   NaN  
251  johndoe@email.com   NaN  
257                NaN   NaN  
264                NaN   NaN  
269                NaN   NaN  
277  johndoe@email.com   NaN  
278                NaN   NaN  
286                NaN   NaN  
296                NaN   NaN

In [84]:
patients_clean.iloc[215:216,:]

patient_id assigned_sex given_name surname          address      city  \
215         216         male       John     Doe  123 Main Street  New York   

    state zip_code        country  birthdate  weight  height   bmi  \
215    NY    12345  United States 1975-01-01   180.0      72  24.4   

                 email phone  
215  johndoe@email.com   NaN

####  `treatments` 表格两列中的三个变量 (治疗、初始剂量和最终剂量)

##### 定义
*治疗类型是一个变量，剂量变化是两个变量，都需要单独成列，用melt方法将治疗类型分组，再删掉空行（即重复的行）。提示：使用 pandas 的 [融合函数](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) 和 [`str.split()` 方法](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html)。这是非常不错的 [`融合` 教程](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

##### 代码

In [35]:
# 这里是你的清理代码
treat = treatments_clean.melt(id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'], value_vars=['auralin', 'novodra'], value_name='treat')
treat = treat[treat.treat != '-']

In [36]:
treat['treat_start'], treat['treat_end'] = treat.treat.str.split(' - ', 1).str

In [37]:
treatments_clean = treat.drop('treat', axis=1)

##### 测试

In [38]:
# 这里是你的测试代码
treatments_clean.sample(5)

given_name    surname  hba1c_start  hba1c_end  hba1c_change variable  \
253       albino  schiavone         7.56       7.15          0.41  auralin   
293     bernarda    cindrić         7.89       7.55          0.34  auralin   
30   szilveszter      totth         7.70       7.38          0.32  auralin   
393         jens    poulsen         7.55       7.16          0.39  novodra   
52            mỹ      quynh         7.61       7.16          0.45  auralin   

    treat_start treat_end  
253         35u       43u  
293         40u       49u  
30          35u       39u  
393         36u       32u  
52          57u       64u

#### 不良反应应是 `treatments` 表格的一部分

##### 定义
*不良反应也属于治疗的一部分。用merge链接起来。提示：解答中使用的函数 [教程 ](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) 。*

##### 代码

In [39]:
# 这里是你的清理代码
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean, on=['given_name', 'surname'], how='outer')

##### 测试

In [40]:
# 这里是你的测试代码
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change variable  \
0   veronika    jindrová         7.63       7.20          0.43  auralin   
1       skye  gormanston         7.97       7.62          0.35  auralin   
2     sophia      haugen         7.65       7.27          0.38  auralin   
3      eddie      archer         7.89       7.55          0.34  auralin   
4       asia     woźniak         7.76       7.37          0.39  auralin   

  treat_start treat_end adverse_reaction  
0         41u       48u              NaN  
1         33u       36u              NaN  
2         37u       42u              NaN  
3         31u       38u              NaN  
4         30u       36u              NaN

####  `patients` 表格中的姓和名两列复制到 `treatments` 和 `adverse_reactions`小写 表格中，以及小写形式的姓和名 

##### 定义
*三个表中姓名的大小写不一样在连接表查询时会造成问题，所以我们把patients里的姓名全部小写，跟另外两个表连接起来，只留下id。提示：解答中使用的函数 [教程](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) 和另一个函数的[教程](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) 。*

##### 代码

In [41]:
# 这里是你的清理代码
name = patients_clean[['patient_id', 'given_name', 'surname']]
name.given_name = name.given_name.str.lower()
name.surname = name.surname.str.lower()

G:\Program Files\Anaconda3\lib\site-packages\pandas\core\generic.py:4405: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [42]:
id_treat = pd.merge(name, treatments_clean, on=['given_name', 'surname'])
treatments_clean = id_treat.drop(['given_name', 'surname'], axis=1)
id_adver = pd.merge(name, adverse_reactions_clean, on=['given_name', 'surname'])
adverse_reactions_clean = id_adver.drop(['given_name', 'surname'], axis=1)

##### 测试

In [43]:
# 这里是你的测试代码
treatments_clean.sample()

patient_id  hba1c_start  hba1c_end  hba1c_change variable treat_start  \
335         482         8.51       8.08          0.43  novodra         40u   

    treat_end adverse_reaction  
335       39u              NaN

In [44]:
adverse_reactions_clean.sample()

patient_id           adverse_reaction
16         255  injection site discomfort

### 质量

<font color='red'>观看 *"清理之质量问题"* 视频后，完成剩余 "质量" **定义、代码和测试** 顺序。</font>

#### 邮编是浮点，而不是字符串，有时是四位数

##### 定义
*邮编应当是字符串而不是浮点数。提示：查看 "数据清理过程" 页面。*

##### 代码

In [45]:
patients_clean['zip_code'][patients_clean.zip_code.isnull() == True] = '00000'

G:\Program Files\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [46]:
# 这里是你的清理代码
patients_clean.zip_code = patients_clean.zip_code.apply(lambda x:'{0:>5d}'.format(int(x)).replace(' ', '0'))

##### 测试

In [47]:
# 这里是你的测试代码
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state zip_code        country  birthdate  weight  \
0  Rancho California  California    92390  United States  7/10/1976   121.7   
1          Armstrong    Illinois    61812  United States   4/3/1967   118.8   
2               York    Nebraska    68467  United States  2/19/1980   177.8   
3         Woodbridge          NJ    07095  United States  7/26/1951   220.9   
4             Dothan          AL    36303  United States  2/18/1928   192.3   

   height   bmi                      email              phone  
0      66  19.6   ZoeWellish@superrito.com       951-719-9170  
1      66  19.2       PamelaSHill@cuvox.de  +1 (217) 569-3204  
2      71  24.8       JaeMDebord@gustr.com       402-363-6804  
3      70  31.7  PhanBaLiem@jourrapide.com  +1 (732) 636-8246  
4      27  26.1        TimNeudorf@cuvox.de       334-515-7487

#### Tim Neudorf 的身高是 27 英尺，而不是 72 英尺

##### 定义
*这里是你的定义。*

##### 代码

In [48]:
# 这里是你的清理代码
patients_clean.height = patients_clean.height.replace(27,72)

##### 测试

In [49]:
# 这里是你的测试代码
patients_clean[patients_clean['given_name'] == 'Tim']

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      72  26.1   

                 email         phone  
4  TimNeudorf@cuvox.de  334-515-7487

#### 有时使用州的全称，其他用简称

##### 定义
*这里是你的定义。提示：解答中使用的函数 [教程](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) 。*

##### 代码

In [50]:
# 这里是你的清理代码
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}
def re_state(x):
    if x in state_abbrev.keys():
        state = state_abbrev[x]
        return state
    else:
        return x
            
patients_clean['state'] = patients_clean['state'].apply(re_state)

##### 测试

In [51]:
# 这里是你的测试代码
patients_clean.state.value_counts(5)
# state_abbrev['California']

CA    0.122200
NY    0.095723
TX    0.065173
IL    0.048880
FL    0.044807
MA    0.044807
PA    0.036660
GA    0.030550
OH    0.028513
LA    0.026477
OK    0.026477
MI    0.026477
NJ    0.024440
VA    0.022403
WI    0.020367
MS    0.020367
TN    0.018330
AL    0.018330
IN    0.018330
MN    0.018330
KY    0.016293
NC    0.016293
WA    0.016293
MO    0.014257
ID    0.012220
NV    0.012220
KS    0.012220
NE    0.012220
IA    0.010183
SC    0.010183
CT    0.010183
ME    0.008147
ND    0.008147
RI    0.008147
CO    0.008147
AR    0.008147
AZ    0.008147
SD    0.006110
OR    0.006110
DE    0.006110
WV    0.006110
MD    0.006110
DC    0.004073
MT    0.004073
VT    0.004073
WY    0.002037
NM    0.002037
AK    0.002037
NH    0.002037
Name: state, dtype: float64

#### Dsvid Gustafsson

##### 定义
*Dsvid的名字应该是David。*

##### 代码

In [54]:
# 这里是你的清理代码
patients_clean['given_name'] = patients_clean['given_name'].replace('Dsvid', 'David')

##### 测试

In [56]:
# 这里是你的测试代码
patients_clean[patients_clean['given_name'] == 'David']

patient_id assigned_sex given_name     surname               address  \
8             9         male      David  Gustafsson    1790 Nutter Street   
280         281         male      David    Beauvais    98 Freshour Circle   
440         441         male      David   Villadsen  1434 Oak Ridge Drive   

             city state zip_code        country   birthdate  weight  height  \
8     Kansas City    MO    64105  United States    3/6/1937   163.9      66   
280   San Antonio    TX    78258  United States    5/1/1931   128.5      67   
440  St Elizabeth    MO    65559  United States  11/30/1924   137.3      68   

      bmi                        email              phone  
8    26.5  DavidGustafsson@armyspy.com       816-265-9578  
280  20.1     DavidLBeauvais@rhyta.com       210-218-3477  
440  20.9  DavidMVilladsen@armyspy.com  +1 (573) 493-4748

#### 错误的数据类型 (指定性别、州、邮编和出生日期列) 和错误的数据类型 (auralin 和 novodra 列)，以及 Auralin 和 Novodra 初始剂量和最终剂量中的字母 'u' 

##### 定义
*性别和州是固定个数的，所以指定类型，邮编是str类型，出生日期是date类的。提示：解答中使用的方法 [文档页面](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html)、函数 [文档页面](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) 以及另一种方法的 [文档页面](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html)。*

##### 代码

In [65]:
# 这里是你的清理代码
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate, errors='coerce')

In [72]:
treatments_clean.treat_start = treatments_clean.treat_start.str.strip('u').astype('int64')
treatments_clean.treat_end = treatments_clean.treat_end.str.strip('u').astype('int64')

##### 测试

In [85]:
# 这里是你的测试代码
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        503 non-null object
country         491 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
email           491 non-null object
phone           485 non-null object
dtypes: category(2), datetime64[ns](1), float64(2), int64(2), object(8)
memory usage: 53.9+ KB


In [86]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
patient_id          349 non-null int64
hba1c_start         349 non-null float64
hba1c_end           349 non-null float64
hba1c_change        349 non-null float64
variable            349 non-null object
treat_start         349 non-null int64
treat_end           349 non-null int64
adverse_reaction    35 non-null object
dtypes: float64(3), int64(3), object(2)
memory usage: 24.5+ KB


#### 多个手机号格式

##### 定义
*把所有符号删掉。提示：有用的 [Stack Overflow 答案](https://stackoverflow.com/a/123681)。*

##### 代码

In [122]:
# 这里是你的清理代码
patients_clean.phone = patients_clean.phone.str.replace(r'[\W]', '')

##### 测试

In [123]:
# 这里是你的测试代码
patients_clean.sample(5)

patient_id assigned_sex given_name    surname                address  \
235         236       female    Justyna  Kowalczyk        3115 May Street   
345         346       female      Asuna     Morita      1324 Duffy Street   
266         267       female    Johanna     Dreher     3857 Straford Park   
116         117         male    Philipp    Ziegler  3686 Meadowcrest Lane   
179         180         male    Dominik  Grunewald    3574 Park Boulevard   

             city state zip_code        country  birthdate  weight  height  \
235  Pippa Passes    KY    41844  United States 1981-04-05   112.2      61   
345   Medaryville    IN    47957  United States 1969-12-30   134.2      61   
266    Whitesburg    KY    41858  United States 1948-10-06   187.0      61   
116         Allen    KY    41601  United States 1941-03-03   143.7      69   
179  Marshalltown    IA    50158  United States 1935-11-20   143.0      71   

      bmi                            email       phone  
235  21.2  JustynaKowalczyk@jourrapide.com  6063689825  
345  25.4        AsunaMorita@superrito.com  2198433924  
266  35.3      JohannaDreher@superrito.com  6066324327  
116  21.2     PhilippZiegler@superrito.com  6064174332  
179  19.9        DominikGrunewald@cuvox.de  6417535678

#### John Doe 的默认数据

##### 定义
*把所有John Doe假数据删掉。记住，假设替换后的  John Doe 数据不可恢复。*

##### 代码

In [131]:
# 这里是你的清理代码
patients_clean = patients_clean[patients_clean.surname != 'Doe']

##### 测试

In [132]:
# 这里是你的测试代码
patients_clean[patients_clean.surname == 'Doe']

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, email, phone]
Index: []

#### Jakobsen、Gersten、Taylor 多条记录

##### 定义
*删除重复的病人记录。*

##### 代码

In [164]:
# 这里是你的清理代码
patients_clean = patients_clean[~(patients_clean.duplicated('address') & patients_clean.address.notnull())]

##### Test

In [165]:
# 这里是你的测试代码
patients_clean[patients_clean.duplicated('address') & patients_clean.address.notnull()]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, email, phone]
Index: []

#### Zaitseva 体重单位是 kgs，不是 lbs 

##### 定义
*把Zaitseva的体重单位从lbs改为kgs，换算公式是2.2046*kg = lb。*

##### 代码

In [175]:
# 这里是你的清理代码
zai_weight = patients_clean[patients_clean.surname == 'Zaitseva'].weight
zai_weight = zai_weight*2.2046
patients_clean.loc[patients_clean.surname == 'Zaitseva', 'weight'] = zai_weight

##### 测试

In [176]:
# 这里是你的测试代码
patients_clean[patients_clean.surname == 'Zaitseva']

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate     weight  height  \
210  Wooster    OH    44691  United States 1938-11-26  107.58448      63   

      bmi                          email       phone  
210  19.1  CamillaZaitseva@superrito.com  3302022145